### ISSUE TITLE
Output JUnit Test Results for Integration into CI Systems

### ISSUE THREAD

**Issue by willbuckner** [koalaman/shellcheck#358]
Hello,

Absolutely love your tool; we use it as part of our automated test suites that run every time code is checked in. The testing tools for our other (non-shell) codebases all generate JUnit output for test failures. I believe it would be a powerful feature to write out a JUnit/xUnit XML file showing failures identified by shellcheck. A good resource on the format is available here: https://stackoverflow.com/questions/4922867/junit-xml-format-specification-that-hudson-supports

If someone that knows Haskell starts it, maybe I can finish it, but I'm not at all familiar with Haskell at the moment :( My only way to generate JUnit output would be from parsing the output from the tool via shell tools. Supporting this would allow simple integration into Bamboo, Hudson, Jenkins, and the like.

**Comment by koalaman** [comment]
Glad you like it!

How do you imagine linting comments mapping to unit test results? I.e., what do you consider to be the classname, name, type and details for output like "Line 4, char 26, code SC2086: Double quote to prevent globbing and word splitting." ?

ShellCheck can already output CheckStyle compatible XML, which many of these types of tools already supports either natively or through plugins.

**Comment by deryni** [comment]
I had just been thinking about this actually (and had just found https://bitbucket.org/atlassian/bamboo-checkstyle-plugin/overview which I had not yet had time to look into trying) so I took a look at the checkstyle output formatter and figured I could probably modify that to output something that looked like JUnit instead.

So I tried my hand at it and came up with this: https://github.com/koalaman/shellcheck/pull/361 .

As indicated in the PR comment I don't know Haskell at all and haven't worked with JUnit before so I have no idea if I got either of those parts at all correct (or correct enough in the JUnit case).

**Comment by koalaman** [comment]
Awesome! 

What I'm wondering though is whether this would be better solved by one of:

* Using the CheckStyle support already present in these tools. For example, the Jenkins CheckStyle plugin mentions a "failure threshold to mark a build as unstable" which I imagine would be similar to having a build with a failing unit test. 

* Writing a JUnit test that just executes `shellcheck` and asserts that the exit status is 0. This would be about 10 lines of Java, and result in a bona fide test failure if that's helpful for the workflow. 

* XSLT, so that you can pipe `shellcheck -f checkstyle` through `xmlstarlet` or `xsltproc checkstyle2junit.xsl` and get a JUnit XML file out. This would be a general and flexible adapter that would work just as well for shellcheck as for jshint or csslint.

**Comment by deryni** [comment]
Yeah. I'm not entirely convinced this is the correct approach either.

I don't know how the various CI/etc. tools consume the input from JUnit or CheckStyle at all so I can't say if using CheckStyle will be better for the workflow or not but it certainly might be. At the very least I would imagine something along those lines would work better with CheckStyle (i.e. a single shellcheck failure doesn't necessarily need or want to flag the entire build as "red").

I don't do Java so can't do the JUnit test concept.

The XSLT option should be easy enough though. I may give that a shot later.

**Comment by deryni** [comment]
Just for the record (since I happened to notice this today). [luacheck](https://github.com/mpeterv/luacheck) uses [this bit of code](https://github.com/mpeterv/luacheck/blob/master/src/luacheck/format.lua#L212) for its JUnit output formatting.

**Comment by aw-altiscale** [comment]
FWIW, we are in the process of integrating shellcheck into Hadoop's patch precommit (we have a _lot_ of bash code).  But junit would definitely be handier.  checkstyle is great for primarily cosmetic things, but at least in our community, it tends not to get run by devs as part of their normal test cycle whereas all of our unit testing is done with Junit and definitely gets run on a regular basis.

**Comment by koalaman** [comment]
I added [checkstyle2junit.xslt](https://github.com/koalaman/shellcheck/wiki/JUnit) to the wiki. Let me know if this helps.

**Comment by Chogos** [comment]
When I used the provided checkstyle2junit.xslt only one failure per file was reported since the resulting XML is not (entirely) valid for the JUnit plugin on Jenkins. I've adjusted it to create a failed testcase for each error which enabled Jenkins to report all of them.

The adjusted script:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">

  <xsl:output encoding="UTF-8" method="xml"></xsl:output>

  <xsl:template match="node() | @*">
    <xsl:copy>
      <xsl:apply-templates select="node() | @*"/>
    </xsl:copy>
  </xsl:template>

  <xsl:template match="/">
    <testsuite>
      <xsl:attribute name="tests">
        <xsl:value-of select="count(.//file)" />
      </xsl:attribute>
      <xsl:attribute name="failures">
        <xsl:value-of select="count(.//error)" />
      </xsl:attribute>
      <xsl:for-each select="//checkstyle">
        <xsl:apply-templates />
      </xsl:for-each>
    </testsuite>
  </xsl:template>

  <xsl:template match="file">
    <xsl:for-each select=".//error">
      <testcase>
        <xsl:attribute name="classname">
          <xsl:value-of select="../@name" />
        </xsl:attribute>
        <xsl:attribute name="name">
          <xsl:value-of select="../@name" />
        </xsl:attribute>
        <failure>
          <xsl:attribute name="type">
            <xsl:value-of select="@source" />
          </xsl:attribute>
          <xsl:text>Line </xsl:text>
          <xsl:value-of select="@line" />
          <xsl:text>: </xsl:text>
          <xsl:value-of select="@message" />
        </failure>
      </testcase>
    </xsl:for-each>
  </xsl:template>

</xsl:stylesheet>
```

**Comment by josegonzalez** [comment]
Just popping in to say that the Dokku project - which uses shellcheck quite a bit, thanks @koalaman and others! - will be using a python script to munge the original shellcheck output. That is available [here](https://github.com/dokku/dokku/pull/3370/files#diff-03eeb0399377a14615a2af43fa0e532a). A few notes:

- The linked to XSLT in the wiki works, but is missing a bunch of data. Worse, it doesn't pass the _only_ xsd I could find online for Junit ([here](https://github.com/windyroad/JUnit-Schema). Good attempt though!
- The XSLT from @Chogos sometimes bailed on my failure output.
- The python script doesn't include passing files. I'd love it if there could be a flag to list all passing files in the original output.
- None of these methods allow the original information to be output _as is_. That is to say, you'd need to run shellcheck twice to get human readable output with xmlstarlet, and my version doesn't support colors. In theory, you could use `tee` with my python script, but I don't want to fiddle with redirection in make targets just now.

One thing that [I did for bats-core](https://github.com/bats-core/bats-core/pull/153) and might make sense here is to have a `raw` output that is then passed through a second command for extra formatting. The raw output could have all the data anyone could want, making it quite easy for people to write formatters for whatever output they like. Supporting multiple formatters would allow both human and machine-readable output, making usage in common CI platforms easier to get into. I'd love to help here, but my haskell is _definitely_ weaker than I'd like it to be.

Regardless, thanks all for your help!
